src/CompositeRheologies.jl

```diff
@@ -0,0 +1,794 @@
+# This holds structures and computational routines for compositional rheologies
+using StaticArrays
+using Setfield
+
+export CompositeRheology, Parallel, create_rheology_string, print_rheology_matrix
+export time_τII_0D, compute_εII_harmonic, compute_τII_AD
+export computeViscosity_εII, computeViscosity_εII_AD
+import Base.getindex
+
+import GeoParams.Units: nondimensionalize, dimensionalize
+
+"""
+    Put rheological elements in parallel 
+"""
+struct Parallel{T, N} <: AbstractConstitutiveLaw{T}
+    elements::T
+end
+
+function Parallel(v::T) where T
+    v = tuple(v...)
+    return Parallel{typeof(v),length(v)}(v)
+end
+Parallel(a,b...) = Parallel((a,b...,)) 
+
+
+@generated function getindex(p::Parallel{T, N}, I::Int64) where {T,N}
+    quote
+        Base.@_inline_meta
+        Base.Cartesian.@nexprs $N i -> I == i && return p.elements[i]
+    end
+end
+
+
+"""
+    Structure that holds composite rheologies (e.g., visco-elasto-viscoplastic),
+    but also indicates (in the name) whether we need to perform non-linear iterations.
+"""
+struct CompositeRheology{T, N, Npar, is_parallel, τ_it, P_it, λ_it} <: AbstractComposite
+    elements::T
+end
+
+# Defines tuples of composite rheologies, while also checking which type of iterations need to be performed
+function CompositeRheology(v::T) where {T, N}
+
+    # These three variables will indicate later which type of non-linear iterations are required
+    τ_it = false;
+    P_it = false;
+    λ_it = false;
+
+    id_parallel =   findall(isa.(v, Parallel));
+    Npar        =   length(id_parallel)
+    n           =   length(v)
+    par         =   zeros(Bool,n)
+    if Npar>0
+        par[id_parallel] .= 1
+    end
+    is_parallel =    SVector{n,Bool}(par)
+
+    return CompositeRheology{typeof(v), n, Npar, is_parallel, τ_it, P_it, λ_it}(v)
+end
+CompositeRheology(a,b...) = CompositeRheology( (a,b...,)) 
+CompositeRheology(a::Parallel) = CompositeRheology( (a,)) 
+
+@generated function getindex(p::CompositeRheology{T, N}, I::Int64) where {T,N}
+    quote
+        Base.@_inline_meta
+        Base.Cartesian.@nexprs $N i -> I == i && return p.elements[i]
+    end
+end
+
+# define rules to nondimensionalise this 
+function nondimensionalize(MatParam::Union{Parallel,CompositeRheology}, g::GeoUnits{TYPE}) where {TYPE}
```

Comment:
Can this function and the one below be moved where all the other methods for `nondimensionalize()` and `dimensionalize()` are?

Reply:
I tried doing that, but it complained about not knowing the `CompositeRheology` and `Parallel` structs. If you know how to fix that, it would be great!

Reply:
No, you are right. It's best to leave it like this for now given how everything is organised.